hw/ip/aes/doc/_index.md

```diff
@@ -344,7 +348,7 @@ Then for every Data Block `I=0,..,N-3`, software must:
    Each register must be read at least once.
    The order in which these registers are read does not matter.
 3. Write Input Data Block `I+2` into the Input Data register.
-   There is no need to check INPUT_READY as the cycle following OUTPUT_VALID being set the current input is loaded in.
+   There is no need to explicitly check INPUT_READY as in the same cycle OUTPUT_VALID becomes `1`, the current input is loaded in (meaning INPUT_READY becomes `1` one cycle later).
```

Comment:
I guess we probably still want to check it, as we don't want to overwrite the input register data. I guess the new value won't be loaded to the internal state until all rounds have finished and the result was written out into the output register. 

So if we have some loop that writes the data into the input data register 3 time sin a row, and does these 3 writes in less than (12/14/16 cycles - depends on the key size), we would "loose" one write to the input data register?

Or is my understanding not correct?

Reply:
I don't think so. If you follow the referenced sequence, you are writing at most 2 input blocks (each 4 regs DATA_IN0 - DATA_IN3) in a row (but even then check the INPUT_READY in between the two blocks).

After that, you check for OUTPUT_VALID, read the 4 output regs and provide the next input block to the four input regs. Here you don't need to check INPUT_READY in between because you know that after OUTPUT_VALID goes high, the module directly continues loading the input into internal state registers. The input register can safely be overwritten without loosing anything.

The reason for doing this is that you safe 1 bus transaction out of 10 per block (4 data writes, 4 data reads, 2 or 1 status reads). Without a DMA, this is the performance bottleneck for AES.